Mathlib/LinearAlgebra/Matrix/Permanent.lean

```diff
@@ -0,0 +1,85 @@
+/-
+Copyright (c) 2024 Moritz Firsching. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching
+-/
+import Mathlib.Algebra.BigOperators.GroupWithZero.Finset
+import Mathlib.Data.Fintype.Perm
+import Mathlib.Data.Matrix.Diagonal
+/-!
+# Permanent of a matrix
+
+This file defines the permanent of a matrix, `Matrix.permanent`, and some of its properties.
+
+## Main definitions
+
+* `Matrix.permanent`: the permanent of a square matrix, as a sum over permutations
+
+-/
+
+open Equiv Fintype Finset
+
+namespace Matrix
+
+variable {n : Type*} [DecidableEq n] [Fintype n]
+variable {R : Type*} [CommSemiring R]
+
+/-- The permanent of a matrix defined as a sum over all permutations -/
```

Comment:
Please make the docstring useful for people who don't know what's the permanent of a matrix. Same for the module docstring. Also, Wikipedia says that some authors define permanent of a rectangular matrix too. If you don't want to do it (I never used permanent in my research, so I have no opinion here), then please mention this choice in the module docstring.

Reply:
Otherwise LGTM. Thanks!
bors d+

Reply:
Optional: add
- `permanent_smul_col`, `permanent_smul_row`, `permanent_smul`
- add `permanent` as a `MultilinearMap`. Do we have symmetric multilinear maps?

Reply:
I added a sentence explaining the relation to determinant; this way the docstring is more useful for people who already know what a determinant is. Also adding `square` in order to make sure that this is not the more exotic more general rectangular permanent. 

Regarding the theorems/definitions you mention: I agree they should be added, but let's do it in another PR.

Reply:
> Optional (can wait for a future PR): add
> 
> * `permanent_smul_col`, `permanent_smul_row`, `permanent_smul`

Adding those in #19307